### ISSUE TITLE
EIP-1193 compatible providers

### ISSUE THREAD

**Issue by alcuadrado** [xf00f/web3x#12]
Hi @xf00f,

The [EIP-1193](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1193.md#L10) defines a standard javascript/typescript Ethereum provider API. It has already been implemented by MetaMask and Mist Browser, and [is being implemented by web3.js](https://github.com/ethereum/web3.js/tree/ethereumProvider).

The simpler way to implement it in web3x is having a special provider that wraps an EIP-1193 compatible one, exposing [this interface](https://github.com/xf00f/web3x/blob/master/src/providers/index.ts#L36,L43). Users of web3x would have to use it in this way:

```js
import { EthereumProviderAdapter } from 'web3x-es/providers';
import { Eth } from 'web3x-es/eth';

const provider = new EthereumProviderAdapter(window.ethereum);
const eth = Eth.fromProvider(provider);
```

From the user perspective it would be better to have native support, using it like this:

```js
import { Eth } from 'web3x-es/eth';

const eth = Eth.fromProvider(window.ethereum);
```

This alternative feels more natural, specially when integrating with MetaMask or other dapp browsers. It may require a major refactor though.

Do you have any thought about? Any plan to implement it?

**Comment by xf00f** [comment]
@alcuadrado I'm aware of this and agree with your analysis. I would prefer to do a proper refactor to support EIP-1193 interface in main codebase, getting rid of RequestManager etc. It may then be possible to modify the http/ws/ipc providers to also support this interface, or write an adapter around them.

I may be able to look into this in next couple of weeks. I'm not sure how much work it is but I guess the tricky bit will be around all the subscription/notification stuff.

**Comment by alcuadrado** [comment]
@xf00f great! I also think that getting rid of `RequestManager` is a good idea, I never understood the need for that class. I may be missing something though.

In the meantime, we'll probably implement wrappers to "transform" a web3x `HttpProvider` into an EIP-1193 one, and an EIP-1193 one into a web3x `Provider`. I'm happy to build them as part of this project, but I they wouldn't add much value once native EIP-1193 support is added.

**Comment by xf00f** [comment]
This has been merged and released as part of v2.0.0.

It was a big refactor. All tests are passing but it's possible there has been a regression. Let me know if you come across anything.

**Comment by alcuadrado** [comment]
Thanks for implementing it this quickly! We are upgrading web3x in Buidler right now :)
